libtransmission/transmission.h

```diff
@@ -829,8 +829,16 @@ tr_torrent* tr_torrentNew(tr_ctor* ctor, tr_torrent** setme_duplicate_of);
 
 using tr_fileFunc = bool (*)(char const* filename, void* user_data, tr_error* error);
 
+using tr_SuccessNotifyFunc = void (*)(bool succeeded, void* user_data);
```

Comment:
I believe it's customary to include resource-identifying information (torrent id in this case, as the whole torrent just won't be available at that point) in callback parameters. If added here, you no longer need `CallbackUserData` and could just use a pointer to `Session` (or `Session::Impl`, whatever is more convenient) as user data, which would be more straight-forward.

Also, please follow the naming convention (casing), and maybe consider replacing "success" with something more conservative, e.g. `tr_torrent_remove_done_func` (to mirror `tr_torrent_rename_done_func`).

Also CC @ckerr: I notice that this file is no longer pure C, given the use of `using` outside of `#ifdef __cplusplus` (that's not a C feature yet, right? :-P); do you want to drop C support altogether (#4568…) and just use `std::function`, at least right here and right now?

Reply:
@mikedld I have been putting off that "is pure C the goal of the public API" decision for a long time, do we have to decide it here?? :bomb:

Reply:
> do we have to decide it here??

I don't think I saw any pushback on C++, despite you making an effort to involve people in the discussion, apart from our own considerations around Swift, and for Swift I guess it doesn't really matter if this C abstraction layer will be right in libransmission or in the Mac client; the latter may even end up being more efficient and convenient anyway since it won't be bound by API stability promises.

Reply:
One thing that might be important in this decision is that C has stable ABI: doesn't matter the version or maker of the compiler, the binary C interface is standard. Not true for C++.

Reply:
We're not distributing libtransmission separately from the clients (I did try to support its isolated installation in the past but I'd be surprised to know that any packagers in fact make use of this, if it's even still working), and those who use it outside of this project will probably build it themselves and link statically anyway, which relaxes the requirements quite a bit.

Reply:
Note for the Mac client, my current approach will be an "ObjC abstraction" header. So that will be an header OK for Swift's Mac client, but hardly usable as a C header for anything else (since it will be ObjC).

Reply:
About the name, it is better to make it consistent with `tr_fileFunc`, which is also used in the funcion`tr_torrentRemove()` along with the new type. That is why I named it `tr_successNotifyFunc`, which follows `tr_fileFunc` in making a reference to _what_ it is used for, not _where_ it is used. I am fixing the uppercase `s`, though.

About passing the torrent id, it won't solve the need to create a struct, as I need to pass the Session pointer and the result `bool` to `g_idle_add()` (or whatever Gtkmm alternative available for triggering a call in the main thread from another thread (which I think there isn't, at least not passing arguments, as `Glib::Dispatcher` doesn't allow them)).

Reply:
For the name, you're _not_ notifying about successful completion of the operation but rather about completion regardless of it being success or failure, so the `tr_successNotifyFunc` name is misleading, in fact violating your suggested "_what_ it is used for" reference (not sure I understand the second part about _where_).

As for the casing, it's my understanding that all new code follows snake_case for identifiers, thus not adding to future work to unify the naming style of preexisting identifiers.

Hope my other comment about `connect_once()` resolves the question around the necessity of `CallbackUserData`, which I still think shouldn't be necessary.

Reply:
I am changing to "tr_result_notify_func". I didn't touch the other the other identifiers to not overload this PR.